src/mlpack/methods/decision_tree/mse_gain.hpp

```diff
@@ -96,54 +96,179 @@ class MSEGain
+
+      for (size_t i = 0; i < minimum - 1; ++i)
+      {
+        const WType w = weights[i];
+        const RType x = responses[i];
+
+        // Calculating initial weighted mean of responses for the left child.
+        leftSize += w;
+        leftMean += w * x;
+        sumSquares[i] = prevWeightedSumSquares + w * x * x;
+        prevWeightedSumSquares += w * x * x;
+      }
+      if (leftSize > 1e-9)
+        leftMean /= leftSize;
+
+      for(size_t i = minimum - 1; i < responses.n_elem; ++i)
+      {
+        const WType w = weights[i];
+        const RType x = responses[i];
+
+        // Calculating initial weighted mean of responses for the right child.
+        rightSize += w;
+        rightMean += w * x;
+        sumSquares[i] = prevWeightedSumSquares + w * x * x;
+        prevWeightedSumSquares += w * x * x;
+      }
+      if (rightSize > 1e-9)
+        rightMean /= rightSize;
+
+      totalSumSquares = prevWeightedSumSquares;
+    }
+    else
+    {
+      // Stores the sum of squares till the previous index.
+      double prevSumSquares = 0.0;
+
+      for (size_t i = 0; i < minimum - 1; ++i)
+      {
+        const RType x = responses[i];
+
+        // Calculating the initial mean of responses for the left child.
+        ++leftSize;
+        leftMean += x;
+        sumSquares[i] = prevSumSquares + x * x;
+        prevSumSquares += x * x;
+      }
+      if (leftSize > 1e-9)
+        leftMean /= leftSize;
+
+      for(size_t i = minimum - 1; i < responses.n_elem; ++i)
+      {
+        const RType x = responses[i];
+
+        // Calculating the initial mean of responses for the right child.
+        ++rightSize;
+        rightMean += x;
+        sumSquares[i] = prevSumSquares + x * x;
+        prevSumSquares += x * x;
+      }
+      if (rightSize > 1e-9)
+        rightMean /= rightSize;
+
+      totalSumSquares = prevSumSquares;
+    }
+  }
+
+  /**
+   * Updates the statistics for the given index.
    *
-   * @param sumSquares Precomputed sum of square (sum(Xi^2)) of values.
-   * @param mean Precomputed mean (sum(Xi) / n) of values.
-   * @param childSize The total number of samples in that child.
+   * @param responses The set of responses on which statistics are computed.
+   * @param weights The set of weights associated to each response.
+   * @param index The current index.
    */
-  static double Evaluate(const double sumSquares,
-                         const double mean,
-                         const size_t childSize)
+  template<bool UseWeights, typename ResponsesType, typename WeightVecType>
+  void UpdateStatistics(const ResponsesType& responses,
+                        const WeightVecType& weights,
+                        const size_t index)
   {
-    double mse = sumSquares / (double) childSize - mean * mean;
-    return -mse;
+    typedef typename ResponsesType::elem_type RType;
+    typedef typename WeightVecType::elem_type WType;
+
+    if (UseWeights)
+    {
+      const WType w = weights[index];
```

Comment:
If the user does not pass `index` as 1 greater than the previous `index` that was used when `UpdateStatistics()` was called, this could give an incorrect result.  I wonder if it might be better to internally store `lastIndex` (initialized to `0`), and loop over all values between `lastIndex + 1` and `index` (inclusive) to update the value, instead of just `index`.

Reply:
I think it is just fine because we don't expect the user to call this function directly.

Moreover, iterating in this way allows us to check if the data value has changed from the last index or not. This helps us to skip computing gain for some of the indexes where the value doesn't change. Now, I don't deny that this can't be done with the way you are suggesting here but for that too, something similar would be needed to skip some indexes. So, I think this is okay the way it is now. Let me know if it doesn't make sense :)

Reply:
Yeah, agreed, there are a couple ways to do it---either you have two functions, one of which takes a step but doesn't return the gain, and the other of which takes a step and returns the gain (this is the way you have it now), or you have one function that allows 'fast forwarding', e.g., taking possibly multiple steps and returning the gain of the last one.

(By 'step' there I mean 'increase the index'.)

Up to you which you want to go with---personally, I think just one function is cleaner, but, I can see advantages and disadvantages to both approaches.  (They are pretty minor tradeoffs though.)

Reply:
I think we can have a single function. The  latter approach that you mentioned seems nice 👍